### DISCUSSION TITLE
Batch replace markdown documents / マークダウン文書の一括置換

### DISCUSSION THREAD

**Discussion by yuki-takei** [growilabs/growi#7180] [Help / Questions] [answered]
Workaround for:

- [Upgrading to v6.0.x](https://docs.growi.org/en/admin-guide/upgrading/60x.html)
- [Upgrading to v6.1.x](https://docs.growi.org/en/admin-guide/upgrading/61x.html)
- [Upgrading to v7.0.x](https://docs.growi.org/en/admin-guide/upgrading/70x.html)

**Comment by yuki-takei** [comment]
WIP: Example script for mongo shell

https://github.com/weseek/growi/pull/7266

**Comment by yuki-takei** [comment]
## Example (WIP)
```
git clone https://github.com/weseek/growi
cd growi/bin/data-migrations/v6

NETWORK=growi_devcontainer_default \
MONGO_URI=mongodb://growi_devcontainer_mongo_1/growi \
docker run --rm \
  --network $NETWORK \
  -v "$(pwd)"/src:/opt \
  -w /opt \
  -e MIGRATION_TYPE=v6 \
  mongo:6.0 \
  /bin/mongosh $MONGO_URI migration.js
```

**Comment by yukendev** [comment, marked as answer]
(日本語のマニュアルは下の方にあります)

# Migration of page body (English)

> **Warning**
> **Migration in this way is applied only to the latest revision. Past revisions are not applied.**

## Usage
```
git clone https://github.com/weseek/growi
cd growi/bin/data-migrations

NETWORK=growi_devcontainer_default \
MONGO_URI=mongodb://growi_devcontainer_mongo_1/growi \
docker run --rm \
  --network $NETWORK \
  -v "$(pwd)"/src:/opt \
  -w /opt \
  -e MIGRATION_MODULE=v60x \
  mongo:6.0 \
  /bin/mongosh $MONGO_URI index.js
```

## Variables
| Variable              | Description                                                                    | Default |
| --------------------- | ------------------------------------------------------------------------------ | ------- |
| NETWORK     | Network in docker compose of MongoDB server                                                         | -       |
| MONGO_URI| URI that can connect to MongoDB                                                     | -       |

## Environment variables
### Required

| Variable              | Description                                                                    | Default |
| --------------------- | ------------------------------------------------------------------------------ | ------- |
| MIGRATION_MODULE     | Specify migration module                                                       | -       |

The value of `MIGRATION_MODULE` is one of the following.
- `v60x/drawio`: Migration for Draw.io notation only([
reference](https://docs.growi.org/en/admin-guide/upgrading/60x.html#specification-change-draw-io-diagrams-net-notation))
- `v60x/plantuml`: Migration for PlantUML notation only([
reference](https://docs.growi.org/en/admin-guide/upgrading/60x.html#specification-change-plantuml-notation))
- `v60x/tsv`: Migration for table notation by TSV only([
reference](https://docs.growi.org/en/admin-guide/upgrading/60x.html#specification-change-notation-for-drawing-tables-with-csv-and-tsv))
- `v60x/csv`: Migration for table notation by CSV only([
reference](https://docs.growi.org/en/admin-guide/upgrading/60x.html#specification-change-notation-for-drawing-tables-with-csv-and-tsv))
- `v60x/bracketlink`: Migration for only page links within GROWI([
reference](https://docs.growi.org/en/admin-guide/upgrading/60x.html#deprecated-growi-original-notation-page-link))
- `v60x` or `v60x/index`: Migration for all notations in v6.0.x series
- `v61x/mdcont`: Migration for mdcont notation only([reference](https://docs.growi.org/en/admin-guide/upgrading/61x.html#specification-change-elimination-of-mdcont-prefix-automatically-added-to-anchor-links))
- `v61x` or `v61x/index`: Migration for all notations in v6.1.x series
- `v70x/bootstrap5`: Migration for Bootstrap4 to Bootstrap5 
- `v70x` or `v70x/index`: Migration for all notations in v7.0.x series
- `custom`: You can define your own processors and apply them to `revision` (see "Advanced" below for details)

### Optional

| Variable              | Description                                                                    | Default |
| --------------------- | ------------------------------------------------------------------------------ | ------- |
| BATCH_SIZE     | Number of revisions to be processed at one time（revision）                                                         | 100       |
| BATCH_INTERVAL| Interval after batch processing（ms）                                                     | 3000       |

※The `BATCH_INTERVAL` is for server load control. If you don't mind the load of the MongoDB, there is no problem to reduce it.

## Advanced

By creating a function in `growi/bin/data-migrations/src/migrations/custom.js` that replaces a specific regular expression, you can replace all specific strings in the latest revisions for all pages.

The following function replaces the string `foo` with the string `bar`.

``` javascript
module.exports = [
  (body) => {
      var fooRegExp = /foo/g; // foo regex
      return body.replace(fooRegExp, 'bar'); // replace to bar
  },
];
```

By passing `custom` in the environment variable `MIGRATION_MODULE` and executing it, you can apply the `custom.js` to all the latest `revisions`.
```
git clone https://github.com/weseek/growi
cd growi/bin/data-migrations

NETWORK=growi_devcontainer_default \
MONGO_URI=mongodb://growi_devcontainer_mongo_1/growi \
docker run --rm \
  --network $NETWORK \
  -v "$(pwd)"/src:/opt \
  -w /opt \
  -e MIGRATION_MODULE=custom \
  mongo:6.0 \
  /bin/mongosh $MONGO_URI index.js
```


# ページ本文のマイグレーション (日本語)

> **Warning**
> **この方法でマイグレーションが適用されるのは最新のrevisionのみです。過去のrevisionには適用されません。**

## 利用例
```
git clone https://github.com/weseek/growi
cd growi/bin/data-migrations

NETWORK=growi_devcontainer_default \
MONGO_URI=mongodb://growi_devcontainer_mongo_1/growi \
docker run --rm \
  --network $NETWORK \
  -v "$(pwd)"/src:/opt \
  -w /opt \
  -e MIGRATION_MODULE=v60x \
  mongo:6.0 \
  /bin/mongosh $MONGO_URI index.js
```

## 必要な変数
| Variable              | Description                                                                    | Default |
| --------------------- | ------------------------------------------------------------------------------ | ------- |
| NETWORK     | MongoDBサーバーのdocker compose内ネットワーク                                                         | -       |
| MONGO_URI| MongoDBへの接続可能なURI                                                     | -       |

## 環境変数
### Required

| Variable              | Description                                                                    | Default |
| --------------------- | ------------------------------------------------------------------------------ | ------- |
| MIGRATION_MODULE     | マイグレーションモジュール名                                                      | -       |

`MIGRATION_MODULE`の値は以下のうちのどれかです。
- `v60x/drawio`: Draw.ioの記法のみをマイグレーション([参考](https://docs.growi.org/ja/admin-guide/upgrading/60x.html#%E4%BB%95%E6%A7%98%E5%A4%89%E6%9B%B4-draw-io-diagrams-net-%E8%A8%98%E6%B3%95))
- `v60x/plantuml`: PlantUMLの記法のみをマイグレーション([参考](https://docs.growi.org/ja/admin-guide/upgrading/60x.html#%E4%BB%95%E6%A7%98%E5%A4%89%E6%9B%B4-plantuml-%E8%A8%98%E6%B3%95))
- `v60x/tsv`: TSVによるテーブル作成の記法のみをマイグレーション([参考](https://docs.growi.org/ja/admin-guide/upgrading/60x.html#%E4%BB%95%E6%A7%98%E5%A4%89%E6%9B%B4-csv-tsv-%E3%81%AB%E3%82%88%E3%82%8B%E3%83%86%E3%83%BC%E3%83%95%E3%82%99%E3%83%AB%E6%8F%8F%E7%94%BB%E8%A8%98%E6%B3%95))
- `v60x/csv`: CSVによるテーブル作成の記法のみをマイグレーション([参考](https://docs.growi.org/ja/admin-guide/upgrading/60x.html#%E4%BB%95%E6%A7%98%E5%A4%89%E6%9B%B4-csv-tsv-%E3%81%AB%E3%82%88%E3%82%8B%E3%83%86%E3%83%BC%E3%83%95%E3%82%99%E3%83%AB%E6%8F%8F%E7%94%BB%E8%A8%98%E6%B3%95))
- `v60x/bracketlink`: GROWI内のページリンクのみをマイグレーション([参考](https://docs.growi.org/ja/admin-guide/upgrading/60x.html#%E6%9C%AA%E5%AE%9F%E8%A3%85-%E5%BB%83%E6%AD%A2%E6%A4%9C%E8%A8%8E%E4%B8%AD-growi-%E7%8B%AC%E8%87%AA%E8%A8%98%E6%B3%95%E3%81%AE%E3%83%98%E3%82%9A%E3%83%BC%E3%82%B7%E3%82%99%E3%83%AA%E3%83%B3%E3%82%AF))
- `v60x` or `v60x/index`:  v6.0.x 系の全てのマイグレーション
- `v61x/mdcont`: header の mdcont prefix のみのマイグレーション([reference](https://docs.growi.org/ja/admin-guide/upgrading/61x.html#%E4%BB%95%E6%A7%98%E5%A4%89%E6%9B%B4-%E3%82%A2%E3%83%B3%E3%82%AB%E3%83%BC%E3%83%AA%E3%83%B3%E3%82%AF%E3%81%AB%E8%87%AA%E5%8B%95%E4%BB%98%E4%B8%8E%E3%81%95%E3%82%8C%E3%82%8B-mdcont-%E3%83%95%E3%82%9A%E3%83%AC%E3%83%95%E3%82%A3%E3%82%AF%E3%82%B9%E3%81%AE%E5%BB%83%E6%AD%A2))
- `v61x` or `v61x/index`: v6.1.x 系の全てのマイグレーション
- `v70x/bootstrap5`: Bootstrap4 記法から Bootstrap5 記法へのマイグレーション
- `v70x` or `v70x/index`:v7.0.x 系の全てのマイグレーション
- `custom`: 自分でプロセッサを定義し、`revision`に適用させることができる(詳細は下の「応用」を参照 )

### Optional

| Variable              | Description                                                                    | Default |
| --------------------- | ------------------------------------------------------------------------------ | ------- |
| BATCH_SIZE     | 1回に処理するrevisionの数（revision）                                                         | 100       |
| BATCH_INTERVAL| バッチ処理後のインターバル（ms）                                                     | 3000       |

※`BATCH_INTERVAL`はサーバーの負荷対策です。実行対象のMongoDBが複数のGROWIで使用されていたり、GROWIが落ちたりすると困る状況ではない場合は、短くしても問題はありません。

## 応用

`growi/bin/data-migrations/src/migrations/custom.js` の中に特定の正規表現を置き換える関数を作成することで、すべての最新の`revision`の特定の文字列を全て置き換えることができます。

以下の関数では、`hogehoge`という文字列を`fugafuga`という文字列に置き換えています。
``` javascript
module.exports = [
  (body) => {
       var fooRegExp = /foo/g; // foo regex
       return body.replace(fooRegExp, 'bar'); // replace to bar
  },
];
```

環境変数の`MIGRATION_MODULE`に`custom`を渡して実行することで、`custom.js` を全ての最新の`revision`に適用させることができます。
```
git clone https://github.com/weseek/growi
cd growi/bin/data-migrations

NETWORK=growi_devcontainer_default \
MONGO_URI=mongodb://growi_devcontainer_mongo_1/growi \
docker run --rm \
  --network $NETWORK \
  -v "$(pwd)"/src:/opt \
  -w /opt \
  -e MIGRATION_MODULE=custom \
  mongo:6.0 \
  /bin/mongosh $MONGO_URI index.js
```

**Reply by miya** [reply]
マイグレーションスクリプトの実行手順が変更されたので更新された手順 ↑ をご確認ください。(2023.5.25)
更新差分はスレッドの右上から確認できます。

**Comment by yuki-takei** [comment]
仕様変更になってしまうけれど、「`MIGRATION_TYPE` を**渡さずに**」よりは、明示的に `custom` みたいなのを渡した方がわかりやすそう。何も渡ってこなかった場合はエラー表示。

**Comment by yuki-takei** [comment]
BATCH_INTERVAL に関しては但し書きを付けよう。

- 負荷対策である
- もし実行対象の MongoDB を他者と共用していないなら短くしてもよい
